### PR TITLE
feat: introduce async rebalance mode for expert load balancer

### PR DESCRIPTION
**Pull request by TheBasy**
- Add CLI arg `--eplb-rebalance-async-mode`
- Background thread: broadcast `logical_count` → compute `ExpertLocationMetadata` → store in `_rebalance_result`
- TP barrier: new gloo `cpu_group` signals (`send_signal_cpu` / `recv_signal_cpu`) ensure all ranks swap plan atomically
- Yield-based generator keeps training loop non-blocking; transfer starts after TP-wide agreement via `_begin_transfer`
- Sync mode (`async=False`) unchanged: blocking single-thread rebalance

## Checklist

- [ ] Format your code according to the [Code Formatting with Pre-Commit](https://docs.sglang.ai/references/contribution_guide.html#code-formatting-with-pre-commit).
- [ ] Add unit tests as outlined in the [Running Unit Tests](https://docs.sglang.ai/references/contribution_guide.html#running-unit-tests-adding-to-ci).
- [ ] Update documentation / docstrings / example tutorials as needed, according to [Writing Documentation](https://docs.sglang.ai/references/contribution_guide.html#writing-documentation-running-docs-ci).
- [ ] Provide throughput / latency benchmark results and accuracy evaluation results as needed, according to [Benchmark and Profiling](https://docs.sglang.ai/references/benchmark_and_profiling.html) and [Accuracy Results](https://docs.sglang.ai/references/accuracy_evaluation.html).
- [ ] For reviewers: If you haven't made any contributions to this PR and are only assisting with merging the main branch, please remove yourself as a co-author when merging the PR.
- [ ] Please feel free to join our Slack channel at https://slack.sglang.ai to discuss your PR.

### CODE REVIEW

**Comment by TheBasy** [comment]
We conducted the bench_serving test with a batch size of 4096 and a request rate of 4. 
The synchronous EPLB mode results are as follows:
```text
============ Serving Benchmark Result ============
Backend:                                 sglang
Traffic request rate:                    4.0
Max request concurrency:                 4096
Successful requests:                     4096
Benchmark duration (s):                  1082.51
Total input tokens:                      16777216
Total generated tokens:                  4194304
Total generated tokens (retokenized):    4179213
Request throughput (req/s):              3.78
Input token throughput (tok/s):          15498.40
Output token throughput (tok/s):         3874.60
Total token throughput (tok/s):          19373.00
Concurrency:                             275.71
----------------End-to-End Latency----------------
Mean E2E Latency (ms):                   72865.09
Median E2E Latency (ms):                 72389.68
---------------Time to First Token----------------
Mean TTFT (ms):                          1113.59
Median TTFT (ms):                        765.44
P99 TTFT (ms):                           6863.33
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          70.14
Median TPOT (ms):                        69.98
P90 TPOT (ms):                           73.48
P99 TPOT (ms):                           75.28
---------------Inter-Token Latency----------------
Mean ITL (ms):                           70.14
Median ITL (ms):                         66.65
P90 ITL (ms):                            79.06
P95 ITL (ms):                            88.69
P99 ITL (ms):                            139.56
Max ITL (ms):                            12072.54
==================================================
```

The asynchronous EPLB mode results are as follows:
```text
============ Serving Benchmark Result ============
Backend:                                 sglang
Traffic request rate:                    4.0
Max request concurrency:                 4096
Successful requests:                     4096
Benchmark duration (s):                  1079.86
Total input tokens:                      16777216
Total generated tokens:                  4194304
Total generated tokens (retokenized):    4179411
Request throughput (req/s):              3.79
Input token throughput (tok/s):          15536.40
Output token throughput (tok/s):         3884.10
Total token throughput (tok/s):          19420.50
Concurrency:                             269.38
----------------End-to-End Latency----------------
Mean E2E Latency (ms):                   71019.86
Median E2E Latency (ms):                 70991.12
---------------Time to First Token----------------
Mean TTFT (ms):                          1079.06
Median TTFT (ms):                        730.95
P99 TTFT (ms):                           6248.78
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          68.37
Median TPOT (ms):                        68.62
P90 TPOT (ms):                           71.70
P99 TPOT (ms):                           72.50
---------------Inter-Token Latency----------------
Mean ITL (ms):                           68.37
Median ITL (ms):                         66.29
P90 ITL (ms):                            77.96
P95 ITL (ms):                            87.58
P99 ITL (ms):                            136.42
Max ITL (ms):                            8238.64
==================================================
```
With the same load (batch 4096, request rate 4 req/s), switching from synchronous to asynchronous EPLB yields a small but consistent throughput gain (~0.3 % more tokens and requests per second) while simultaneously cutting latency: mean end-to-end time drops ~2-3 %, P99 TTFT and max inter-token latency fall 9 % and 32 %, respectively, and the average in-flight concurrency shrinks by ~2 %. These improvements confirm that moving expert-parameter transfers to a background thread effectively overlaps communication with computation, giving higher hardware utilization and noticeably better tail-latency for large-batch, high-concurrency inference workloads.

File: python/sglang/srt/distributed/parallel_state.py [hunk 1]
```diff
@@ -814,6 +814,17 @@ def recv_object(self, src: int) -> Any:
 
         return obj
 
+    def send_signal_cpu(self, dst_rank_global: int, value: int, group: Optional[ProcessGroup] = None,) -> None:
```
Comment (fzyzcjy) [inline python/sglang/srt/distributed/parallel_state.py]: usually we do not add code here

File: python/sglang/srt/eplb/eplb_manager.py [hunk 2]
```diff
@@ -18,6 +19,12 @@ def __init__(self, model_runner: "ModelRunner"):
         super().__init__()
         self._model_runner = model_runner
         self._server_args = model_runner.server_args
+        self._compute_ongoing = False
+        self._tp_sync_ongoing = False
+        self._i = 0
```
Comment (fzyzcjy) [inline python/sglang/srt/eplb/eplb_manager.py]: would be great to cleanup the code a bit, e.g. here

File: python/sglang/srt/eplb/eplb_manager.py [hunk 3]
```diff
@@ -40,31 +47,127 @@ def __init__(self, model_runner: "ModelRunner"):
...
+        yield
+        while self._tp_sync_ongoing:
+            yield 
+        max_tensor = torch.stack(buffer_kept).max()
+        max_value = max(max_tensor.item(),self._send_sig_i)
+        self._begin_transfer = max_value + 2
+
+    def _compute_expert_metadata(self, logical_count):
+        local_rank = self._model_runner.tp_rank
+        torch.cuda.set_device(local_rank%8)
```
Comment (fzyzcjy) [inline python/sglang/srt/eplb/eplb_manager.py]: e.g. here

**Review by fzyzcjy — COMMENTED** [review]
oh I realize the code is not ready, thus ignore the code cleanup suggestions below for now - as long as the code is clean and neat when it is ready then it is good!

File: python/sglang/srt/distributed/parallel_state.py (hunk 1 above)
Comment (TheBasy) [reply python/sglang/srt/distributed/parallel_state.py]: done

File: python/sglang/srt/eplb/eplb_manager.py (hunk 2 above)
Comment (TheBasy) [reply python/sglang/srt/eplb/eplb_manager.py]: done

File: python/sglang/srt/eplb/eplb_manager.py (hunk 3 above)
Comment (TheBasy) [reply python/sglang/srt/eplb/eplb_manager.py]: done

**Comment by TheBasy** [comment]
> oh I realize the code is not ready, thus ignore the code cleanup suggestions below for now - as long as the code is clean and neat when it is ready then it is good!

Thanks for the heads-up!
The code is now complete and clean, so feel free to take a fresh look whenever convenient.
Really appreciate your time and earlier suggestions! @fzyzcjy

File: python/sglang/srt/eplb/eplb_manager.py [hunk 4]
```diff
@@ -40,31 +47,157 @@ def __init__(self, model_runner: "ModelRunner"):
...
+        num_gpu_per_node = self._num_gpu_per_node
+        torch.cuda.set_device(local_rank % num_gpu_per_node)
         expert_location_metadata = ExpertLocationMetadata.init_by_eplb(
             self._server_args, self._model_runner.model_config, logical_count
         )
-
+        self._rebalance_result = expert_location_metadata
+        self._compute_ongoing = False
+
+    def tp_sync(self):
```
Comment (fzyzcjy) [inline python/sglang/srt/eplb/eplb_manager.py:133]: hmm not sure whether this is optimal or not (do not have time to dig deeper). but FYI there is a PollBasedBarrier that may look somehow related

Comment (TheBasy) [reply python/sglang/srt/eplb/eplb_manager.py:133]: Thanks for the note!
In our tp_sync function, we're aiming to implement a non-blocking all-reduce max operation to synchronize the step_counter across TP ranks. The goal is to ensure that all TP ranks proceed to the data transfer phase only after the maximum step_counter among them is reached, while allowing each rank to continue decoding independently during the synchronization window.
If TP ranks enter the transfer phase at different decoding steps, the earlier ones would have to transfer one layer and then wait for others to finish their current decode — this creates global blocking and hurts throughput.
We did try using torch.distributed.all_reduce directly, but it blocks the main inference thread due to the collective communication being synchronous, which disrupts the continuous decoding flow. That’s why we’re exploring alternative designs to achieve asynchronous coordination without stalling inference.

**Comment by harleyszhang** [comment]
what moe model you are test? can you should give your model server command, that is good for test reproduction.

**Comment by TheBasy** [comment]
> what moe model you are test? can you should give your model server command, that is good for test reproduction.

I test Deepseek-R1.  You only need to add one line, --enable-eplb-rebalance-async, to the original launch arguments to enable asynchronous computation during rebalancing. It is also recommended to set --eplb-rebalance-layers-per-chunk 1, which allows layerwise data migration during the rebalancing phase, further reducing the user's perception of interruption latency.

for example, my original launch arguments for the decode stage is:
```bash
python3 -m sglang.launch_server \
    --disaggregation-mode decode \
    --disaggregation-transfer-backend mooncake \
    --attention-backend flashmla \
    --trust-remote-code \
    --enable-dp-attention \
    --mem-fraction-static 0.9 \
    --context-length 65535 \
    --decode-log-interval 50 \
    --enable-cache-report \
    --moe-dense-tp-size 1 \
    --enable-dp-lm-head \
    --enable-deepep-moe \
    --deepep-mode low_latency \
    --disaggregation-ib-device mlx5_bond_0,mlx5_bond_1,mlx5_bond_2,mlx5_bond_3 \
    --model-path /home/models/deepseek-ai__DeepSeek-R1 \
    --host 0.0.0.0 \
    --port ${DEFAULT_PORT} \
    --dist-init-addr ${DECODE_MASTER_IP}:${MASTER_PORT} \
    --tp-size $((8 * NUM_DECODE)) \
    --dp-size $((DP_SIZE_PER_DECODE_NODE * NUM_DECODE)) \
    --chunked-prefill-size $((DP_SIZE_PER_DECODE_NODE * NUM_DECODE * CHUNKED_PREFILL_SIZE_PER_DP_RANK)) \
    --max-running-requests $((DP_SIZE_PER_DECODE_NODE * NUM_DECODE * DECODE_MAX_RUNNING_REQUEST_PER_DP_RANK)) \
    --nnodes ${NUM_DECODE} \
    --node-rank $node_rank \
    --enable-dp-attention \
    --enable-expert-distribution-metrics \
    --enable-eplb \
    --eplb-rebalance-num-iterations 1000 \
    --cuda-graph-max-bs ${DECODE_MAX_RUNNING_REQUEST_PER_DP_RANK} \
    > ${WORKDIR}/decode.log 2>&1 &
```
After using async rebalance mode, the launch arguments will be
```bash
python3 -m sglang.launch_server \
    --disaggregation-mode decode \
    --disaggregation-transfer-backend mooncake \
    --attention-backend flashmla \
    --trust-remote-code \
    --enable-dp-attention \
    --mem-fraction-static 0.9 \
    --context-length 65535 \
    --decode-log-interval 50 \
    --enable-cache-report \
    --moe-dense-tp-size 1 \
    --enable-dp-lm-head \
    --enable-deepep-moe \
    --deepep-mode low_latency \
    --disaggregation-ib-device mlx5_bond_0,mlx5_bond_1,mlx5_bond_2,mlx5_bond_3 \
    --model-path /home/models/deepseek-ai__DeepSeek-R1 \
    --host 0.0.0.0 \
    --port ${DEFAULT_PORT} \
    --dist-init-addr ${DECODE_MASTER_IP}:${MASTER_PORT} \
    --tp-size $((8 * NUM_DECODE)) \
    --dp-size $((DP_SIZE_PER_DECODE_NODE * NUM_DECODE)) \
    --chunked-prefill-size $((DP_SIZE_PER_DECODE_NODE * NUM_DECODE * CHUNKED_PREFILL_SIZE_PER_DP_RANK)) \
    --max-running-requests $((DP_SIZE_PER_DECODE_NODE * NUM_DECODE * DECODE_MAX_RUNNING_REQUEST_PER_DP_RANK)) \
    --nnodes ${NUM_DECODE} \
    --node-rank $node_rank \
    --enable-dp-attention \
    --enable-expert-distribution-metrics \
    --enable-eplb \
    --eplb-rebalance-layers-per-chunk 1 \
    --enable-eplb-rebalance-async \
    --eplb-rebalance-num-iterations 1000 \
    --cuda-graph-max-bs ${DECODE_MAX_RUNNING_REQUEST_PER_DP_RANK} \
    > ${WORKDIR}/decode.log 2>&1 &

**Comment by jovany-wang** [comment]
Do you have the related benchmark results only on decode instance?

**Comment by jovany-wang** [comment]
@TheBasy Could you merge the latest master to resolve the conflicts?

**Comment by TheBasy** [comment]
> Quote reply
> Reference in new

You can check this PR for the merge: https://github.com/antgroup/sglang/pull/3. It includes the latest changes from master and resolves the conflicts.
